controllers/storageclassclaimsecret/storageclassclaimsecret_controller.go

```diff
@@ -0,0 +1,665 @@
+	var reconcileError error
+	if storagecluster.IsOCSConsumerMode(r.storageCluster) {
+		return reconcile.Result{}, fmt.Errorf("StorageClassClaim Secret is not supported in OCSConsumer mode")
+	}
+
+	result, reconcileError = r.reconcileProviderPhases()
+
+	claimBytes, marshalErr := json.Marshal(r.storageClassClaim)
+	if marshalErr != nil {
+		r.log.Error(marshalErr, "Failed to marshal StorageClassClaim Secret.")
+		return reconcile.Result{}, marshalErr
+	}
+	r.storageClassClaimSecret.Data["StorageClassClaim"] = claimBytes
+
+	// Apply status changes to the StorageClassClaim
+	statusError := r.Client.Update(r.ctx, r.storageClassClaimSecret)
+	if statusError != nil {
+		r.log.Info("Failed to update StorageClassClaim Secret status.")
+	}
+
+	// Reconcile errors have higher priority than status update errors
+	if reconcileError != nil {
+		return result, reconcileError
+	}
+
+	if statusError != nil {
+		return result, statusError
+	}
+
+	return result, nil
+}
+
+// isClaimSecret determines whether we want to reconcile a given Secret
+func isClaimSecret(obj runtime.Object) bool {
+	secretObj, ok := obj.(*corev1.Secret)
+	if !ok {
+		return false
+	}
+
+	annotations := secretObj.GetAnnotations()
+	if _, found := annotations[v1alpha1.StorageClassClaimAnnotation]; found {
+		return true
+	}
+	return false
+}
+
+var claimPredicate = predicate.Funcs{
+	CreateFunc: func(e event.CreateEvent) bool {
+		return isClaimSecret(e.Object)
+	},
+	DeleteFunc: func(e event.DeleteEvent) bool {
+		return isClaimSecret(e.Object)
+	},
+	UpdateFunc: func(e event.UpdateEvent) bool {
+		return isClaimSecret(e.ObjectNew)
+	},
+	GenericFunc: func(e event.GenericEvent) bool {
+		return isClaimSecret(e.Object)
+	},
+}
+
+func (r *StorageClassClaimSecretReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	enqueueStorageConsumerRequest := handler.EnqueueRequestsFromMapFunc(
+		func(obj client.Object) []reconcile.Request {
+			annotations := obj.GetAnnotations()
+			if annotation, found := annotations[v1alpha1.StorageClassClaimAnnotation]; found {
+				parts := strings.Split(annotation, "/")
+				return []reconcile.Request{{
+					NamespacedName: types.NamespacedName{
+						Namespace: parts[0],
+						Name:      parts[1],
+					},
+				}}
+			}
+			return []reconcile.Request{}
+		})
+	// Compose a predicate that is an OR of the specified predicates
+	//secretsPredicate := util.ComposePredicates(
+	//	predicate.GenerationChangedPredicate{},
+	//	util.MetadataChangedPredicate{},
+	//)
+
+	// As we are not setting the Controller OwnerReference on the ceph
+	// resources we are creating as part of the StorageClassClaim, we need to
+	// set IsController to false to get Reconcile Request of StorageClassClaim
+	// for the owned ceph resources updates.
+	enqueueForNonControllerOwner := &handler.EnqueueRequestForOwner{OwnerType: &v1alpha1.StorageClassClaim{}, IsController: false}
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&corev1.Secret{}, builder.WithPredicates(claimPredicate)).
```

Comment:
I didn't seem to have luck with providing both the `claimPredicate` and `predicate.GenerationChangedPredicate{}`, the Claims didn't get updated from `Creating` to `Ready`. Maybe I didn't wait long enough? All I know is that when I removed the predicate the process completed successfully. The following seems to work:

```diff
@@ -202,17 +202,17 @@ func (r *StorageClassClaimSecretReconciler) SetupWithManager(mgr ctrl.Manager) e
                        return []reconcile.Request{}
                })
        // Compose a predicate that is an OR of the specified predicates
-       //secretsPredicate := util.ComposePredicates(
-       //      predicate.GenerationChangedPredicate{},
-       //      util.MetadataChangedPredicate{},
-       //)
+       secretsPredicate := util.ComposePredicates(
+               predicate.GenerationChangedPredicate{},
+               util.MetadataChangedPredicate{},
+       )
 
        // As we are not setting the Controller OwnerReference on the ceph
        // resources we are creating as part of the StorageClassClaim, we need to
        // set IsController to false to get Reconcile Request of StorageClassClaim
        // for the owned ceph resources updates.
        enqueueForNonControllerOwner := &handler.EnqueueRequestForOwner{OwnerType: &v1alpha1.StorageClassClaim{}, IsController: false}
-       return ctrl.NewControllerManagedBy(mgr).
+       return ctrl.NewControllerManagedBy(mgr).WithEventFilter(secretsPredicate).
                For(&corev1.Secret{}, builder.WithPredicates(claimPredicate)).
                Watches(&source.Kind{Type: &rookCephv1.CephBlockPool{}}, enqueueForNonControllerOwner).
                Watches(&source.Kind{Type: &rookCephv1.CephFilesystemSubVolumeGroup{}}, enqueueForNonControllerOwner).
```

There's a lot of weird stuff happening underneath in the builder functions:

* Using [`util.ComposePredicates`](https://github.com/red-hat-storage/ocs-operator/blob/main/controllers/util/predicates.go#L10-L47) generates a single predicate that is an OR of any predicates you pass into it. We put this together so we could catch any of multiple metadata changes we often (though not always) care about.
* The reason the compose exists is that any slice of predicates provided to `builder.WithPredicates()` generates an AND of all listed predicates, which is not always what we want.

I'm not entirely sure why `claimPredicate AND GenerationChangedPredicate` didn't seem to work, but passing the latter through `WithEventFilter()` got the job done. I'm still working on updating the PR for the other comments, but if you end up responding to this before the next push let me know if you'd accept the above diff.

Reply:
Thanks for reminding me about the AND nature of predicates. But you shouldn't have created your own OR behavior when you have `predicate.Or` implemented by the framework. 

Also, I think `WithEventFilter` is not necessary when using the built-in OR function (or any of the other built-in functions). It will set the predicate on all `Watches` and not only on the `For` watch. 
This can, and will, result in erroneous behavior for other events.

Reply:
You have just saved me so much time, THANK YOU.

Reply:
Done